rmw_fastrtps_cpp/include/rmw_fastrtps_cpp/TypeSupport_impl.h

```diff
@@ -203,13 +253,83 @@ static size_t size_of(const MembersType *members)
                 size = sizeof(uint64_t);
                 break;
             case ::rosidl_typesupport_introspection_cpp::ROS_TYPE_STRING:
-                // StringHelper exposes rosidl_generator_c__String as std::string
-                // so we don't need differentiate between C and C++ introspection typesupport
+                // Note: specialization needed because size_of is called before
+                // casting submembers as rosidl_generator_c__String
                 size = sizeof(std::string);
                 break;
             case ::rosidl_typesupport_introspection_cpp::ROS_TYPE_MESSAGE:
                 {
-                    const MembersType *sub_members = (const MembersType*)last_member.members_->data;
+                    const rosidl_typesupport_introspection_cpp::MessageMembers *sub_members = (const rosidl_typesupport_introspection_cpp::MessageMembers*)last_member.members_->data;
+                    size = size_of(sub_members);
+                }
+                break;
+        }
+
+        if(last_member.is_array_)
+            size *= last_member.array_size_;
+    }
+
+    return last_member.offset_ + size;
+}
+
+// C specialization
```

Comment:
`alignof` does return the same value for both the `rosidl_generator_c__String`  representation and `std::string`. It could be worth it to check if it's the case on other architecture (e.g. ARM)